### ISSUE TITLE
Site Editor: Create more robust autosaving including UI

### ISSUE THREAD

**Issue by annezazu** [WordPress/gutenberg#29577]
## What problem does this address?

[As part of the second FSE outreach call for testing](https://make.wordpress.org/test/2021/02/18/fse-program-testing-call-2-build-a-homepage-with-site-editing-blocks/#comment-1367), someone reported a crashing experiencing yet found that there weren't any autosaves in place. This begs the question of what's currently expected for autosaving especially since FSE includes so many entities. It's unclear to me what's expected here but this is something that users have grown accustomed to so it should be considered. 

## What is your proposed solution?

This is tough to say due to the multi-entity saving process we currently have! Perhaps something the autosave could reflect that with a multi-entity selection option for what to restore.

**Comment by jameskoster** [comment]
I feel like there are several parts to this:

1. Autosaving should work as normal. That is to say, if I edit an entity in the Site Editor, an autosaved draft should be created in the background.
2. The UI could potentially reflect autosaves per entity, and it may be possible to discard those changes _before_ entering the save flow. Or maybe even save changes independently.
3. If I exit without saving changes, when I next return to the Site Editor I should be notified that autosaves exist per entity, with options to restore all, or specific entities.
4. If I exit without saving then edit an entity with autosaves in isolation (like a Template Part), I should see the standard autosave notice we see on posts / pages. 

For that second point, since we now denote the name of global blocks like Template Parts and Reusable Blocks in the Toolbar, and may soon list which Template Parts the current template consumes (#29147), we might include something there. Dots are common patterns to indicate unsaved changes in other apps:

<img width="1050" alt="Screenshot 2021-03-05 at 09 43 29" src="https://user-images.githubusercontent.com/846565/110097709-440a6e00-7d97-11eb-84d0-b885163f54bb.png">

We should be careful with the dot though, since we have in the past used this to indicate when a template has been customised from its stock theme-supplied state.

**Comment by paaljoachim** [comment]
We also need to remember that Reusable blocks now have a similar save structure as template parts. At the moment it is not possible to discard a save. This can be tricky as undoing might not bring back the original reusable block. Forcing the user to either recreate the contents of the reusable block or to just accept whatever changes will be saved writing over the original Reusable block.

**Comment by paaljoachim** [comment]
Looking at what we have today and making some adjustments.
Going for a minimum change.

1-Change the Site Title.
2-Clicking Apply.
3-Unchecking the Header. (Discard changes)
4-Clicking Save.
5-Saved. (Still in the Full Site Editing screen. One is just pushed back into the page content. As it is up to the user to choose to go back to Page content when they are ready to do so.)

https://user-images.githubusercontent.com/5323259/111227956-b2470000-85e3-11eb-84a6-34a0093fe0eb.mp4


Figma prototype:
https://www.figma.com/proto/De3yU46oeTResqpxkRPojb/FSE?node-id=30%3A133&scaling=scale-down-width

**Comment by hedgefield** [comment]
I agree with @jameskoster, an autosave should save a draft for any and all changes. Whether someone wants to discard some of those changes before publishing the draft is up to them. Showing the changes in the pre-publish sidebar like that also makes for a cool changes summary!

Minor thing: for the label in Paal's mockup I would suggest "changes will be lost" or "discarding changes". Now it looks like an action you still have to take, like the "remove image" red link that appears when selecting a featured image.

**Comment by jameskoster** [comment]
Yeah I think it is fair to say that if a user actively un-checks an entity in this flow, they have acknowledged that it will not be saved. We probably don't need to display anything extra.

This is probably a bridge too far at this point, but it will be interesting to see the canvas update based on which entities are checked / unchecked in the save sidebar.

**Comment by paaljoachim** [comment]
Here is another version. This can be used in Full Site Editing and in Reusable blocks.

![FSE-Changes-will-be-lost](https://user-images.githubusercontent.com/5323259/111634590-f4945b00-87f6-11eb-8426-75c8a8d7e1af.jpg)

It would in general for various issues be good to look at what would be a good first version. Then iterate from there.
Pushing forward the very first version while still holding a larger vision in mind.

**Comment by jameskoster** [comment]
I still don't think you need the "Changes will be lost" text. It's basically communicating the same thing as the un-checked checkbox? :)

**Comment by paaljoachim** [comment]
I initially also thought that one could just uncheck for items one does not want to save. But to make it real clear it is nice to also give a message confirmation. Just in case the users mind is wandering a bit too far into something else...:)

**Comment by javierarce** [comment]
I also feel we don't need that text. It's a bit redundant and makes the UI quite noisy (imagine if a user had 5 unchecked changes).

What if instead, we make the text above a little scarier?

> **Select the changes you want to save**
> Some changes may affect other areas of your site.
> Unchecked changes will be lost.

**Comment by paaljoachim** [comment]
Yes! That is even better. Thanks @javierarce Javier!

![FSE-Changes-will-be-lost-next-version](https://user-images.githubusercontent.com/5323259/111768691-8c9e4d00-88a8-11eb-929c-72adbb6d77a4.jpg)

**Comment by javierarce** [comment]
I really liked @jameskoster solution to indicate which areas have been modified and allow to save and discard the changes. Here's a little iteration on his idea. 

In this version, the dots are just indicators without any interactivity and the actions are accessible through an ellipsis icon.

https://user-images.githubusercontent.com/4933/111808463-5080e180-88d4-11eb-8a4a-b72bfd153e84.mov

Here's a [Figma file](https://www.figma.com/file/u4wmgxBLPvNg0vGdROnReS/FSE-Create-more-robust-autosaving-including-UI?node-id=102%3A5742) with this prototype.

**Comment by jameskoster** [comment]
Looking good @javierarce! Using an ellipsis for the actions makes much more sense :) 

I can see why the Sidebar bubbled up to the top of the list when you reverted the Header changes, but I wonder if we need to do that? Is the benefit of displaying unsaved areas at the top of the list worth the trade-off in the lost semantics (IE list placement matching position in the document tree)?

**Comment by javierarce** [comment]
Maybe I'm overdesigning this, but if themes end up having several custom parts, I think that placing the modified areas at the top will make the list easier to read and work with. If the general use case is to have three or four areas, I think keeping the list ordered as you proposed probably makes more sense.

**Comment by javierarce** [comment]
I've just seen [this issue](https://github.com/WordPress/gutenberg/issues/29687) and I was wondering if it would make sense to give the option to revert the template state in the popover.

Template areas without changes would have this one option, while modified templates would allow to Save, Discard, and Revert.

**Comment by jameskoster** [comment]
Yup, I think that makes sense. But it should also be possible to revert a template part while editing it in isolation.

**Comment by paaljoachim** [comment]
Let's take a step back. 
First iteration. Let's make it possible for the user to uncheck what they do not want to save and still be able to click the Save and Publish button if they choose to do so. At the moment unchecking blocks one from moving onward.
I suggest removing "Needs Design" and adding the label "Need Dev"

**Comment by jameskoster** [comment]
I can't make up my mind on whether the areas should jump around when changes are discarded 🙃 Hopefully it's something we can try when a PR has been assembled.

I think the design in https://github.com/WordPress/gutenberg/issues/29577#issuecomment-802958322 is ready for development. But this is blocked by https://github.com/WordPress/gutenberg/issues/29147.

I suspect the autosaving behaviours described [here](https://github.com/WordPress/gutenberg/issues/29577#issuecomment-791302392) will need to be separate.

**Comment by paaljoachim** [comment]
Clicking a dot to accomplish an action is not an obvious action. This is hidden UX. As one can not see that an action is hidden behind a dot. See my comment here: https://github.com/WordPress/gutenberg/issues/29871#issuecomment-811830297

It would be nice to go with an obvious in our face save or discard save method. Atleast as a start. Then one can gradually iterate. 

Such as I show in this comment further above. https://github.com/WordPress/gutenberg/issues/29577#issuecomment-802732673

If the user unchecks both header and footer then the button can change to say "Discard save"

![Discard-Save-button](https://user-images.githubusercontent.com/5323259/113422221-9c4f8280-93cc-11eb-92d1-607777567afb.jpg)

Bottom line is having an easy to see method on discarding a save. Discarding a save the above header and footer would revert back to the last changes that were saved or go back to default state. @critterverse Channing and @jameskoster James.

**Comment by annezazu** [comment]
Quick note that as work evolves with real time collaboration https://github.com/WordPress/gutenberg/issues/52593, we'll need to revisit this if we are going to rely on auto-saving. This begs the question of if we need a "multi-entity" autosave endpoint rather than triggering separate autosaves for each entity.

**Comment by talldan** [comment]
My understanding, any entity that supports revisions supports autosaving (happy to be corrected about it if I'm wrong). That includes templates, template parts, and nav menus.

Existing code in the data package also supports autosaving any entity from what I can see.

The problem is that the post editor is the only one that has an `<AutosaveMonitor />` component implemented, and I expect there's other similar UI needed (like notices), not to mention the in-memory autosave system to make the site editor work.

The multi-entity aspect might cause some issues, but then the post editor has also been using Synced Pattern / Reusable Block entities for a number of years. They're not autosaved AFAIK, but do support revisions. I also think if a user is using revisions, they may already be used to this, if they go back to an old revision it might be using entities that no longer exist or were at a different version. I don't think there's any syncing of revision versions across entities (again correct me if I'm wrong, maybe @ramonjd can correct me on this). For example, if you edit a post and a synced pattern at the same time, and you try to restore the old version of a post, it will always have the latest version of the synced pattern, not the version of the synced pattern that existed at the time.

Given all that and the existing trade-offs that exist, perhaps the best place to start is autosaving templates and seeing how that works.

I think it might also be an idea to try autosaving synced patterns, template parts and other entities, but perhaps only when edited in the isolated editing mode. A lot of that should probably come for free.

**Comment by ramonjd** [comment]
> if you edit a post and a synced pattern at the same time, and you try to restore the old version of a post, it will always have the latest version of the synced pattern

This is how I understand things to work right now too.

I double-checked for my own benefit, so treat all this as my "notes" 😄:

- Revisions in WordPress are per-entity rows in `wp_posts` (`post_type = 'revision'`) keyed via `post_parent`. Restoring a revision only swaps the parent post’s fields with the revision’s fields.
- Synced patterns, template parts, etc., each live in their own custom post types (`wp_template`, `wp_template_part`, `wp_block`) and maintain their revisions independently. In block markup, a synced pattern is serialized as `<!-- wp:block {"ref":123} /-->`, so restoring a post revision just restores that block + attributes, and rendering still pulls the current content for post ID 123.
- As Dan already [mentioned](https://github.com/WordPress/gutenberg/issues/29577#issuecomment-3659101632), `wp_block` [supports revisions](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/post.php#L354), and I see new revision rows whenever I update a synced pattern. Hypothetically we could explore tracking those relationships via post meta or something similar, **but don't quote me on that**. In fact forget I ever said it :trollface:
